src/Raven.Client/Documents/MetadataAsDictionary.cs

```diff
@@ -0,0 +1,145 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using Sparrow.Json;
+
+namespace Raven.Client.Documents
+{
+    public class MetadataAsDictionary : IDictionary<string, string>
+    {
+        private readonly Lazy<IDictionary<string, string>> _lazyMetadata;
```

Comment:
We can probably skip the Lazy here, there is no concurrent access or anything like that.
We can just store a null reference and create it when needed